Fix compiler crash involving pattern match with literal type

som-snytt opened the pull request:
In `case s @ "" => s`, it may be dubious to elide
the binding. With an expected literal type,
the case body was incorrectly `s.type`, so that it was not adapted to constant type. This commit
fixes the latter issue only, especially since
-Wunused does not warn about terms of constant type anyway (because they were probably inlined): `case "" => ""`.

Fixes scala/bug#12127

som-snytt committed:
Don't narrow a constant while stabiliszing

In `case s @ "" => s`, it may be dubious to elide
the binding. With an expected literal type,
the case body was incorrectly `s.type`, so that it
was not adapted to constant type. This commit
fixes the latter issue only, especially since
-Wunused does not warn about terms of constant type anyway
(because they were probably inlined): `case "" => ""`.

som-snytt commented:
The change also enables inlining of the field value in the RHS of an accessor:
```
private[this] val x: Int(42) = 42
def x: Int(42) = 42  // inlined this.x
```
The RHS also gets an OriginalTreeAttachment. There is no change to the generated code.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -641,7 +641,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         narrowIf(checkStable(tree), sym.isModuleNotMethod)
       else if (isModuleTypedExpr)                     // (3)
-        narrowIf(tree, condition = true)
+        narrowIf(tree, condition = !sym.isConstant)
```
Maybe `!isConstantType(tree.tpe)` is somewhat more direct, as `narrowIf` is calling `tree.setType`?

som-snytt replied:
You're probably right. I forget that `tree.symbol.tpe` is different from `tree.tpe`, so I'm not sure why I reached for the symbol test, except that first I added the condition to `isModuleTypedExpr`. `#3` is about the tree type.

som-snytt committed:
Push constant type test into isNarrowable

som-snytt commented:
Rebased and added a commit that moves the constant test into `isStableContext`, which is what flipped when the `pt` changed from `String` to `"hello, world"`. The difference is seen in `isNarrowable` (the technical term for it). The method is used in a few places, so the improvement may ameliorate or deteriorate in other subtle ways.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1055,7 +1045,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           context.deprecationWarning(tree.pos, sym)
         tree match {
-          case Literal(`value`) => tree
+          case Literal(`value`) /*| Bind(_, _)*/ => tree
```
Not obvious whether `case s @ ""` getting squashed is surprising to tooling.

SethTisue commented:
> stabiliszing

is this... Hungarian notation?

som-snytt commented:
@SethTisue https://fosstodon.org/@som_snytt/113306593431459494

lrytz reviewed (approved):
That looks right 👍
